### PR TITLE
Pipeline: use tokenizer pad token at generation time if the model pad token is unset.

### LINKED ISSUE #29378: Misleading warning message about pad_token_id when passing tokenizer instance to pipeline
### System Info

Python=3.11.5
Transformers= '4.37.2'

In my setup I am initializing a tokenizer and want to pass it to the pipeline. My expectation is that if I set the `pad_token_id` directly on tokenizer instance, the pipeline should not print the warning `Setting pad_token_id to eos_token_id:50256 for open-end generation.` If I pass the `pad_token_id` directly to the pipeline's `__call__`, the warning is not printed. However, I would rather like to set the `pad_token_id` directly rather than having to think about everytime I use the `__call__` method of the pipeline.

Alternatively, I would like to add it as a parameter to the pipeline instantiation, but I think passing tokenizer parameters in the pipeline generator is currently not envisioned, as discussed in #12039 #24707 #22995 

@Narsil 

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Here is a minimal code example to demonstrate the difference:
```
from transformers import AutoTokenizer, pipeline, AutoModelForCausalLM, Conversation

msg = 'The capital of France '
modelname = 'microsoft/DialoGPT-small'

# hand over model & tokenizer instantiations to pipeline
model = AutoModelForCausalLM.from_pretrained(modelname)
tokenizer = AutoTokenizer.from_pretrained(modelname)
tokenizer.pad_token_id = tokenizer.eos_token_id

chatbot = pipeline(task='conversational', model=model, tokenizer=tokenizer, framework='pt')
messages = ([{"role": "system", "content": 'You are a helpful assistant'}, {"role": "user", "content": msg}])
response = chatbot(Conversation(messages=messages))
```
This prints the warning message
`Setting 'pad_token_id' to 'eos_token_id':50256 for open-end generation.`

The following code does not:
```
from transformers import AutoTokenizer, pipeline, AutoModelForCausalLM, Conversation

msg = 'The capital of France '
modelname = 'microsoft/DialoGPT-small'

chatbot = pipeline(task='conversational', model=modelname, framework='pt')
messages = ([{"role": "system", "content": 'You are a helpful assistant'}, {"role": "user", "content": msg}])
response = chatbot(Conversation(messages=messages), pad_token_id=chatbot.tokenizer.eos_token_id)
```

### Expected behavior

I would expect the first code snippet not to print the warning as the `pad_token_id` is directly set on the tokenizer instance

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Fixes #29378 

The tagged issue describes the problem, the title describes the fix :D

Example of a script that no longer emits a warning, after this PR:
```py
from transformers import AutoTokenizer, AutoModelForCausalLM, pipeline

model = AutoModelForCausalLM.from_pretrained("openai-community/gpt2")
tokenizer = AutoTokenizer.from_pretrained("openai-community/gpt2")
tokenizer.pad_token_id = tokenizer.eos_token_id

llm = pipeline(task='text-generation', model=model, tokenizer=tokenizer, framework='pt')
response = llm('The capital of France ')
```

### CODE REVIEW

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 1]
```diff
@@ -496,6 +496,10 @@ def _forward(self, model_inputs, return_timestamps=False, generate_kwargs=None):
             else:
                 generate_kwargs["encoder_outputs"] = encoder(inputs, attention_mask=attention_mask)
 
+            # If the tokenizer has a pad token but the model doesn't, we add it to the `generate` call
+            if self.tokenizer.pad_token_id is not None and self.model.generation_config.pad_token_id is None:
```
Comment (amyeroberts) [inline src/transformers/pipelines/automatic_speech_recognition.py:500]: I'm assuming we can assume `generation_config` will have the `pad_token_id` attribute?

Comment (amyeroberts) [inline src/transformers/pipelines/automatic_speech_recognition.py:500]: What's the reason for setting this here - this seems like something we should set on the model's generation config when creating the pipeline. 

If we do that, we can then directly check that the padding token is set as expected when the pipeline is created

File: tests/pipelines/test_pipelines_text_generation.py [hunk 2]
```diff
@@ -477,3 +479,19 @@ def test_pipeline_length_setting_warning(self):
...
+        tokenizer = AutoTokenizer.from_pretrained("hf-internal-testing/tiny-random-gpt2")
+        model.config.pad_token_id = None
+        model.generation_config.pad_token_id = None
+        tokenizer.pad_token_id = tokenizer.eos_token_id
+
+        llm = pipeline(task="text-generation", model=model, tokenizer=tokenizer, framework="pt")
+        with self.assertRaises(AssertionError) as exc:
+            with self.assertLogs("transformers", level="WARNING"):
+                llm("The capital of France ")
+        self.assertIn("no logs of level WARNING or higher triggered", str(exc.exception))
```
Comment (amyeroberts) [inline tests/pipelines/test_pipelines_text_generation.py:497]: This is a very flaky test - essentially if anyone adds a new check which might legitimately raise a warning this is going to fail. 

Admittedly, we should try to avoid having any warnings trigger in our tests (we should run things properly) but it's going to be pretty annoying to try and handle.

I'd rather we just didn't add this test tbh

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this! 

Just some qs about the implementation

File: src/transformers/pipelines/automatic_speech_recognition.py (hunk 1 above)
Comment (gante) [reply src/transformers/pipelines/automatic_speech_recognition.py:500]: Yes :) it is part of [`__init__`](https://github.com/huggingface/transformers/blob/6b660d5ed561cc7326e77e3d833cde8085b23663/src/transformers/generation/configuration_utils.py#L353)

Comment (gante) [reply src/transformers/pipelines/automatic_speech_recognition.py:500]: We can set it at __init__ time, yes. IMO not in generation config (permanently changes the object, = magic pattern unseen to the user), but storing it in `forward_params`. `forward_params` is then passed to `_forward`, which calls `generate`. We can even make it a generic pipeline operation, as opposed to pipeline-specific logic that needs to be added to each pipeline.

There is one case this alternative won't fix, though: if the user decides to set `tokenizer.pad_token_id` _after_ initializing the pipeline, only the current solution can handle it gracefully.

Happy to proceed with the change, or to keep as is. (Slight preference for the change, as it is solution with cleaner code)

Comment (amyeroberts) [reply src/transformers/pipelines/automatic_speech_recognition.py:500]: I like the forward params idea :) 

Since the pipeline is intended to be used as a one-liner way to get results, modifying the tokenizer seems counter to getting it to work. Is this something we commonly ask people to do for the pipelines? If so then probs best to leave as-is

Comment (gante) [reply src/transformers/pipelines/automatic_speech_recognition.py:500]: It's the first time I've seen this issue [tokenizer has a pad token but the model config doesn't], so it is not commonly asked. 

Going to push the change :)

File: src/transformers/pipelines/conversational.py [hunk 3]
```diff
@@ -196,9 +196,7 @@ def new_user_input(self):
     build_pipeline_init_args(has_tokenizer=True),
     r"""
         min_length_for_response (`int`, *optional*, defaults to 32):
-            The minimum length (in number of tokens) for a response.
-        minimum_tokens (`int`, *optional*, defaults to 10):
```
Comment (gante) [inline src/transformers/pipelines/conversational.py:200]: `minimum_tokens` is an unused internal variable, probably a legacy version of `min_length`. 

Initially, I removed it from the signature of the private `_forward`. Then, I realized we can remove all traces, since it is unused :)

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 4]
```diff
@@ -311,14 +311,14 @@ def _sanitize_parameters(
 
         forward_params = defaultdict(dict)
         if max_new_tokens is not None:
-            forward_params["generate_kwargs"]["max_new_tokens"] = max_new_tokens
+            forward_params["max_new_tokens"] = max_new_tokens
```
Comment (gante) [inline src/transformers/pipelines/automatic_speech_recognition.py:314]: Note in this file's diff, also applicable for the diff in `src/transformers/pipelines/image_to_text.py`: 

The conventional strategy to pass kwargs to `generate` is through `**forward_params`. Previously in this file, the generation kwargs were held as `forward_params["generate_kwargs"]`, which prevented the use of the conventional strategy. There isn't really a reason to hold these kwargs separately, `generate` is the only sink for kwargs in models that can generate. Models that can't generate will throw an exception regardless of the container for kwargs. As such, this diff aims at minimizing the difference across pipelines :)

**Comment by gante** [comment]
@amyeroberts ready for a re-review :)

**Review by amyeroberts — COMMENTED** [review]
Thanks for iterating on this! 

Happy with the changes in general but I think we should maintain the clear separation between "generate_kwargs" and forward_kwargs. It makes it easier to understand what controls what in the pipeline

File: src/transformers/pipelines/automatic_speech_recognition.py (hunk 4 above)
Comment (amyeroberts) [reply src/transformers/pipelines/automatic_speech_recognition.py:314]: Actually - I realise what I've said about the forward kwargs is wrong here - we can just assume they're passed to the model. In this case, my preference is to still have "generate_kwargs" explicitly in the forward_kwargs, but I don't feel strongly and don't mind if you leave as-is

**Review by amyeroberts — APPROVED** [review]

File: src/transformers/pipelines/automatic_speech_recognition.py (hunk 4 above)
Comment (gante) [reply src/transformers/pipelines/automatic_speech_recognition.py:314]: I think we can agree that regardless of the pattern we choose here, it should be applied to all pipelines with generative capabilities for consistency. Based on this premise, enforcing a separation of `generate_kwargs` this exact way will break backward compatibility, i.e. the following would not be possible

```py
from transformers import pipeline

llm = pipeline(task='text-generation', model="openai-community/gpt2")
response = llm('The capital of France ', max_length=50)
```

Nevertheless, I am aligned with you -- we should separate them! We can do it through `generation_config.update(**kwargs)`, and perform the required validation with the aid of `generation_config.validate()`. One of the requirements to do so is to have a single big blob of keyword arguments to untangle, and thus these changes go in this direction.

Let me know if you agree, in which case I'll merge the PR and prepare this follow-up. [My instinct was to merge this PR now, but I've held it back -- I've merged too many not-100%-approved PRs recently 😉 ]

Comment (amyeroberts) [reply src/transformers/pipelines/automatic_speech_recognition.py:314]: Yeah, let's merge atm so this is unblocked and then we can iterate on something different :)
